docs/rollbacks.md

```diff
@@ -0,0 +1,30 @@
+# Rollback framework documentation
+
+balenaCloud and balenaOS support [host OS Updates](https://www.balena.io/docs/reference/OS/updates/self-service/) Rollbacks is a framework designed to roll back the OS update in case something goes wrong. There are 2 types of failure states that are currently considered.
+
+## rollback-health
+The new OS gets to userspace but something is unhealthy. Userspace is functional and we can use systemd services and bash scripts in this case.
+- This state is checked by a systemd service: `rollback-health.service`.
+- During a HUP, a flag file `rollback-health-breadcrumb` is left in the state partition to enable the `rollback-health` systemd service on next boot.
+- `rollback-health.service` runs `rollback-health` which runs `rollback-tests`. Two things are checked to establish if balenaOS is healthy or not.
+ - balenaEngine not working. The balenaEngine healthcheck is run.
+  - VPN is not connecting but it used to in the previous OS.
+- These tests are run once every minute for 15 minutes which is the default value of the `ROLLBACK_HEALTH_TIMEOUT` variable.
+- If the OS is considered healthy, `rollback-health` clears the flag files left in the state partition. This service won't run again.
+- If rollback-health is triggered, the previous OS boot hooks are run to restore previous boot files, `resin_root_part` is updated in `resinOS_uEnv.txt` to point to the previous OS partition, a flag file `rollback-health-triggered` is left in the state partition, and a reboot it triggered.
+
+## rollback-altboot
```

Comment:
Not sure about this section, the purpose is hazy for me. When you say "The new OS does not get to userspace." does it mean in essence: "rollback-altboot comes into play, if the device encountered any errors before running the userspace successfully (a kernel panic, ...)?

Just saying "The new OS does not get to userspace." feels dangling on its own.

Maybe this could also be on the top, if I understand things correctly: "There are two rollback mechanisms in the Operating System, covering different update failure modes: one based on health checks (`rollback-health`), and another recognizing if the new system is unbootable for some reason (`rollback-altboot`). Their detailed operations are explained below.". .... (Not dictation, just some brainstorming).